helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -37,14 +39,23 @@
     NODES_CONTAIN_SAME_PARTITION,
   }
 
+  public HardConstraint(Type failureReason) {
+    _type = failureReason;
+  }
+
+  public String getFailureReason() {
+    // by default just return the type, extend the method if more details are needed
+    return _type.toString();
```

Comment:
It might depends on the implementation logic.
We can have a default method with the basic message like this. So let's comment that a child class shall override this method if necessary.